### PR TITLE
[Utility] Use TreeStore as source of truth

### LINKED ISSUE #875: [Utility] Address the discrepancy between the state trees and txIndexer
## Objective

The state trees are the "source of truth" and as such should be used for verification purposes. Currently when attempting to add a transaction to the mempool, a check to see whether it has already been committed to state and indexed is carried out. This is important, however the check is done against the `TxIndexer` not the `Transactions` state tree.

See: `utility/transaction.go`

```diff
// HandleTransaction implements the exposed functionality of the shared utilityModule interface.
func (u *utilityModule) HandleTransaction(txProtoBytes []byte) error {
	txHash := coreTypes.TxHash(txProtoBytes)

	// Is the tx already in the mempool (in memory)?
	if u.mempool.Contains(txHash) {
		return coreTypes.ErrDuplicateTransaction()
	}
-       // Is the tx already committed & indexed (on disk)?
-       if txExists, err := u.GetBus().GetPersistenceModule().TransactionExists(txHash); err != nil {
-               return err
-       } else if txExists {
-               return coreTypes.ErrTransactionAlreadyCommitted()
-       }
+       // Is the tx included in the Transactions state tree?
+       root, nodeStore := u.GetBus().GetTreeStore().GetTree(trees.TransactionTreeName)
+       lazy := smt.ImportSparseMerkleTree(nodeStore, sha256.New(), root)
+        proof, err := lazy.Prove(txHash)
+        if err != nil {
+                return err
+        }
+        txHashBz, err := hex.DecodeString(txHash)
+        if err != nil {
+                return err
+        }
+        if valid := smt.VerifyProof(proof, root, txHashBz, txProtoBytes, lazy.Spec()); valid {
+                return coreTypes.ErrTransactionAlreadyCommitted()
+        }
+        // Validate the Transactions tree was used to calculate the root hash
+        stateHash, nodeStore := u.GetBus().GetTreeStore().GetTree(trees.RootTreeName)
+       lazy := smt.ImportSparseMerkleTree(nodeStore, sha256.New(), root)
+        proof, err := lazy.Prove([]byte(trees.TransactionsTreeName))
+        if err != nil {
+                return err
+        }
+        if valid := smt.VerifyProof(proof, stateHash, []byte(trees.TransactionsTreeName), root, lazy.Spec()); valid {
+                return errors.New("malicious transactions tree provided - not part of previous state hash")
+        }

        // Can the tx be decoded?
        tx := &coreTypes.Transaction{}
@@ -20,6 +38,7 @@ func (u *utilityModule) HandleTransaction(txProtoBytes []byte) error {
                return coreTypes.ErrProtoUnmarshal(err)
        }
        // Does the tx pass basic validation?
	if err := tx.ValidateBasic(); err != nil {
		return err
	}

	// Store the tx in the mempool
	return u.mempool.AddTx(txProtoBytes)
}
```

By using the `Transactions` state tree in `persistence/trees/trees.go` we can guarentee the inclusion or exclusion of any element. There may be particular instances where the `TxIndexer` does not have the correct representation of the values in the state tree and as such may lead to invalid transactions entering the mempool - or not allowing valid transactions to enter the mempool.

By relying on a secondary data structure that is **not** the source of truth we are leading to potential areas where the system can be gamed. The transition towards using the state trees for proofs and verification will result in a stronger and more secure network.

## Origin Document

https://www.notion.so/pocketnetwork/2023-06-30-33d9d44c9432406cb0126a1e40bf8cbf?pvs=4#95d544c38bb14d3d806fb1ea6a043b07

Protocol Hour: 30/06/2023

Screenshot 2023-07-04 at 15 50 49

## Goals

- Remove the dependency on the `TxIndexer` for verification - use this only for data retrieval
- Use the various state trees where they are required for proof generation and verification

## Deliverable

- [ ] Following the logic detailed above implement a helper function to verify the inclusion of a certain transaction in the transactions state tree
  - [ ] Integrate this helper function in the `TransactionExists` method of the persistence module interface
- [ ] Do a **single** check to verify the transactions state tree was used to generate the previous state hash
- [ ] Remove the dependency on the `TxIndexer` where it is more appropriate to use the state trees

## Non-goals / Non-deliverables

- Remove the `TxIndexer` entirely
- Change any state tree logic

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)
- [ ] **k8s LocalNet**: verify a `k8s LocalNet` is still functioning correctly by following the instructions [here](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md)

---

**Creator**: @h5law
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by h5law**
### Summary generated by Reviewpad on 26 Jul 23 11:24 UTC
This pull request includes the following changes:
1. In the `trees.go` file:
   - Added a new method `Prove` that generates and verifies a proof against a tree name stored in the `TreeStore`.
   - Updated the `GetTreeHashes` method to exclude the root tree from the map of tree names to their root hashes.
   
2. In the `treestore_module.go` file:
   - Added a new method `Prove` that generates and verifies a proof against a tree with a matching name.
   - No other changes have been made to the existing methods in the file.

3. In the `transaction_test.go` file:
   - Modified the import statements to use different package names.
   - Modified the test cases `TestHandleTransaction_ErrorAlreadyInMempool`, `TestHandleTransaction_ErrorAlreadyCommitted`, and `TestHandleTransaction_BasicValidation`.
   - Modified the function `prepareEmptyIndexedTransaction`.
   - Changed some variable names.

4. In the `ibc_msg_mempool_test.go` file:
   - Added a new test function `TestHandleMessage_ErrorAlreadyCommitted`.
   - Added environment preparation code, code for indexing a test transaction, and handling the error.
   - Modified the error handling code.

5. In the `utility/transaction.go` file:
   - Added an import statement for the `github.com/pokt-network/pocket/shared/crypto` package.
   - Updated the assignment of the `txHash` variable.
   - Updated the condition for checking if the transaction is already in the mempool.
   - Updated the call to `u.GetBus().GetPersistenceModule().TransactionExists`.
   - Updated the check for the existence of the transaction.

6. In the `persistence/trees/trees_test.go` file:
   - Updated the package imports.
   - Added a new test function `TestTreeStore_Prove` and several test cases.
   - Updated the `treeStore` struct and the existing test function.

7. In the `persistence_module.go` file:
   - Added a new method `TransactionExists` and commented out the old method.

Please let me know if you would like a more detailed review of any specific part of the diff.

## Issue

Fixes #875 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Update `TransactionExists` to use TreeStore
- Add `Prove` method to TreeStore
- Update tests

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: utility/transaction_test.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 
 func TestHandleTransaction_ErrorAlreadyInMempool(t *testing.T) {
 	// Prepare test data
-	emptyTx := types.Transaction{}
+	emptyTx := coreTypes.Transaction{}
```
Comment (Olshansk) [inline utility/transaction_test.go:20]: I feel like you had another PR that renamed it to `core_types`, now? If so, let's stay consistent.

File: shared/modules/treestore_module.go [hunk 2]
```diff
@@ -31,6 +31,9 @@ type TreeStoreModule interface {
 	Update(pgtx pgx.Tx, height uint64) (string, error)
 	// DebugClearAll completely clears the state of the trees. For debugging purposes only.
 	DebugClearAll() error
+	// Prove generates and verifies a proof against the tree with the matching name using the given
+	// key and value. If value == nil, it will verify non-membership of the key, otherwise membership.
+	Prove(treeName string, key, value []byte) error
```
Comment (Olshansk) [inline shared/modules/treestore_module.go:36]: Can you have this return (bool, error) so we can differentiate between errors and just incorrect key-value pairs?

File: persistence/trees/trees.go [hunk 3]
```diff
@@ -103,6 +103,26 @@ func (t *treeStore) GetTree(name string) ([]byte, kvstore.KVStore) {
 	return nil, nil
 }
 
+// Prove generates and verifies a proof against the tree name stored in the TreeStore
+// using the given key-value pair. If value == nil this will be an exclusion proof,
+// otherwise it will be an inclusion proof.
+//
+// If the proof is valid, nil is returned. Otherwise, an error is returned.
```
Comment (Olshansk) [inline persistence/trees/trees.go:110]: I think we should `(bool, error)` to be explicit

File: persistence/block.go [hunk 4]
```diff
@@ -1,33 +1,24 @@
...
-	res, err := p.txIndexer.GetByHash(hash)
-	if res == nil {
-		// check for not found
-		if err != nil && errors.Is(err, badger.ErrKeyNotFound) {
-			return false, nil
-		}
-		return false, err
+func (p *persistenceModule) TransactionExists(txHash, txProtoBz []byte) bool {
+	exists := p.GetBus().GetTreeStore().Prove(trees.TransactionsTreeName, txHash, txProtoBz)
+	if bytes.Equal(txProtoBz, nil) && exists == nil {
```
Comment (Olshansk) [inline persistence/block.go:18]:
```suggestion
	// exclusion verification
	if bytes.Equal(txProtoBz, nil) && exists == nil {
```

File: persistence/block.go [hunk 5]
```diff
@@ -1,33 +1,24 @@
...
+		return false
 	}
-
-	return true, nil
+	return exists == nil
```
Comment (Olshansk) [inline persistence/block.go:21]:
```suggestion
	// inclusion verification
	return exists == nil
```

File: ibc/ibc_msg_mempool_test.go [hunk 6]
```diff
@@ -250,10 +250,39 @@ func TestHandleMessage_ErrorAlreadyInMempool(t *testing.T) {
...
+	require.NoError(t, err)
+	txProtoBytes, err := codec.GetCodec().Marshal(validPruneTx)
+	require.NoError(t, err)
+
+	idxTx := &coreTypes.IndexedTransaction{
+		Tx:            txProtoBytes,
+		Height:        0,
+		Index:         0,
+		ResultCode:    0,
+		Error:         "h5law",
```
Comment (Olshansk) [inline ibc/ibc_msg_mempool_test.go:268]: lol

File: utility/transaction_test.go [hunk 7]
```diff
@@ -37,10 +37,40 @@ func TestHandleTransaction_ErrorAlreadyInMempool(t *testing.T) {
...
+	txProtoBytes, err := codec.GetCodec().Marshal(emptyTx)
+	require.NoError(t, err)
+
+	// Test data - Prepare IndexedTransaction
+	idxTx := &coreTypes.IndexedTransaction{
+		Tx:            txProtoBytes,
+		Height:        0,
+		Index:         0,
+		ResultCode:    0,
+		Error:         "Olshansky",
```
Comment (Olshansk) [inline utility/transaction_test.go:56]: I very much appreciate the strings in this test

File: persistence/trees/trees_test.go [hunk 8]
```diff
@@ -22,3 +30,76 @@ func TestTreeStore_DebugClearAll(t *testing.T) {
...
+
+	testCases := []struct {
+		name        string
+		treeName    string
+		key         []byte
+		value       []byte
+		expectedErr error
+	}{
+		{
+			name:        "valid proof: key and value in tree",
```
Comment (Olshansk) [inline persistence/trees/trees_test.go:59]: Can you update the test names with `inclusion` and `exclusion` so its more semantically self explanatory?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few nits

File: utility/transaction_test.go (hunk 1 above)
Comment (h5law) [reply utility/transaction_test.go:20]: yeah so i saw @bryanchriswhite mention this was a more appropriate way with the underscores. But the file already used camelCase so i left it like that just fixed the issue where it imported core types 2x with diff names. Will make them snake_case

**Review by Olshansk — APPROVED** [review]
flowers-tree

**Pull request merged**